File: AlgorithmsandDataStructures/11thlab/A(SimpleFlow).js
var INF = 1000000;

var n = readline();
var m = readline();
var pipeline = [];
var nums = [];
for (var i = 0; i < n; ++i) {
   nums[i] = [];
 }
for (var i = 0; i < m; ++i) {
  var args = readline().split(" ").map(function(x) {
    return parseInt(x);
  });
  pipeline.push([args[2], --args[0], --args[1], 0]);
  pipeline.push([args[2], args[1], args[0], 0]);
  nums[args[0]].push(2 * i);
  nums[args[1]].push(2 * i + 1);
}

var came = [];
var prev = [];

for (var i = 0; i < n; ++i) {
  came.push(false);
  prev.push(-1);
}

var bfs = function () {
  for (var i = 0; i < n; ++i) {
    came[i] = false;
    prev[i] = -1;
  }
  var q = [];
  q.push(0);
  came[0] = true;
  var flow = INF;
  while (q.length !== 0) {
    var v = q.shift();
    for (var i = 0; i < nums[v].length; ++i) {
      var gap = pipeline[nums[v][i]][0] - pipeline[nums[v][i]][3];
      var pipe = pipeline[nums[v][i]][2];
      if (came[pipe] === false && gap > 0) {
        if (gap < flow) {
          flow = gap;
        }
        q.push(pipe);
        came[pipe] = true;
        prev[pipe] = nums[v][i];
      }
    }
  }
  if (came[n - 1] === true) {
    return flow;
  } else {
    return 0;
  }
};

var dfs = function (v, flow) {
  if (v === n - 1) return flow;
  for (var i = 0; i < nums[v].length; ++i) {
    if (pipeline[nums[v][i]][3] > 0) {
      var volume = dfs(pipeline[nums[v][i]][2],
          Math.min(flow, pipeline[nums[v][i]][3]));
      pipeline[nums[v][i]][3] -= volume;
      return volume;
    }
  }
  return 0;
};

var volume = bfs();
while (volume != 0)
{
  for (var v = n - 1; v != 0; v = pipeline[prev[v]][1]) {
    pipeline[prev[v]][3] += volume;
    pipeline[prev[v] ^ 1][3] -= volume;
  }
  volume = bfs();
}
var volumes = [];
for (var i = 0; i < 2 * m; i += 2) {
  volumes.push(pipeline[i][3]);
}
volume = 0;
var tmp = dfs(0, INF);
while (tmp !== 0) {
  volume += tmp;
  tmp = dfs(0, INF);
}
print(volume);
for (var i = 0; i < volumes.length; ++i) {
  print(volumes[i]);
}

File: AlgorithmsandDataStructures/11thLab/A(SimpleFlow).cpp
#include <iostream>
#include <vector>
#include <queue>

const unsigned int INF = 1e6;

class Pipe
{
	public:
		int volume = 0;
		unsigned int capacity;
		unsigned short inTo, outOf;
		Pipe(unsigned int capacity, unsigned short inTo, unsigned short outOf)
		: capacity(capacity), inTo(inTo), outOf(outOf) {}
};

std::vector <int> prevv;
std::vector <std::vector <unsigned int>> nums;
std::vector<Pipe> pipeline;
unsigned short n;

unsigned int bfs()
{
	std::vector <bool> used(n, 0);
	prevv.assign(n, -1);
	std::queue<unsigned short> q;
	q.push(0);
	used[0] = 1;
	unsigned int flow = INF;
	while (!q.empty())
	{
		for (unsigned int i : nums[q.front()])
		{
			int gap = pipeline[i].capacity - pipeline[i].volume;
			unsigned short pipe = pipeline[i].outOf;
			if (!used[pipe] && gap > 0)
			{
				if (gap < flow)
				{
					flow = gap;
				}
				q.push(pipe);
				used[pipe] = 1;
				prevv[pipe] = i;
			}
		}
		q.pop();
	}
	return used[n - 1] ? flow : 0;
}

int dfs(unsigned short v, int flow)
{
	if (v == n - 1)
	{
		return flow;
	}
	for (unsigned int j : nums[v])
	{
		if (pipeline[j].volume > 0)
		{
			int volume = dfs(pipeline[j].outOf, std::min(flow, pipeline[j].volume));
			pipeline[j].volume -= volume;
			return volume;
		}
	}
	return 0;
}

int main()
{
	unsigned int m;
	std::cin >> n >> m;
	nums.resize(n);
	for (unsigned int i = 0; i < m; ++i)
	{
		unsigned short a, b;
		unsigned int c;
		std::cin >> a >> b >> c;
		pipeline.push_back(Pipe(c, --a, --b));
		pipeline.push_back(Pipe(c, b, a));
		nums[a].push_back(2 * i);
		nums[b].push_back(2 * i + 1);
	}
	unsigned int volume = bfs();
	while (volume)
	{
		for (unsigned short v = n - 1; v != 0; v = pipeline[prevv[v]].inTo)
		{
			pipeline[prevv[v]].volume += volume;
			pipeline[prevv[v] ^ 1].volume -= volume;
		}
		volume = bfs();
	}
	std::vector <int> volumes;
	for (unsigned int i = 0; i < 2 * m; i += 2)
	{
		volumes.push_back(pipeline[i].volume);
	}
	volume = 0;
	int tmp = dfs(0, INF);
	while (tmp)
	{
		volume += tmp;
		tmp = dfs(0, INF);
	}
	std::cout << volume << '\n';
	for (int i : volumes) std::cout << i << '\n';
}
